chainstate/test-suite/src/tests/fungible_tokens_v1.rs

```diff
@@ -0,0 +1,2141 @@
+            .build_and_process();
+
+        assert_eq!(
+            result.unwrap_err(),
+            ChainstateError::ProcessBlockError(BlockError::StateUpdateFailed(
+                ConnectTransactionError::AttemptToPrintMoney(
+                    amount_to_mint,
+                    amount_to_unmint_over_limit,
+                )
+            ))
+        );
+
+        // Unmint some tokens
+        tf.make_block_builder()
+            .add_transaction(
+                TransactionBuilder::new()
+                    .add_input(
+                        TxInput::from_account(
+                            nonce,
+                            AccountSpending::TokenCirculatingSupply(token_id),
+                        ),
+                        InputWitness::NoSignature(None),
+                    )
+                    .add_input(
+                        TxInput::from_utxo(mint_tx_id.into(), 0),
+                        InputWitness::NoSignature(None),
+                    )
+                    .add_input(
+                        TxInput::from_utxo(mint_tx_id.into(), 1),
+                        InputWitness::NoSignature(None),
+                    )
+                    .add_output(TxOutput::Burn(OutputValue::Coin(
+                        token_min_supply_change_fee,
+                    )))
+                    .add_output(TxOutput::Burn(OutputValue::TokenV1(
+                        token_id,
+                        amount_to_unmint,
+                    )))
+                    .build(),
+            )
+            .build_and_process()
+            .unwrap();
+
+        let actual_supply =
+            TokensAccountingStorageRead::get_circulating_supply(&tf.storage, &token_id).unwrap();
+        assert_eq!(
+            actual_supply,
+            Some((amount_to_mint - amount_to_unmint).unwrap())
+        );
+    });
+}
+
+// Check that if supply is unlimited up to i128::MAX tokens can be minted.
+#[rstest]
+#[trace]
+#[case(Seed::from_entropy())]
+fn mint_unlimited_supply(#[case] seed: Seed) {
+    utils::concurrency::model(move || {
+        let mut rng = make_seedable_rng(seed);
+        let mut tf = TestFramework::builder(&mut rng).build();
+
+        let token_min_supply_change_fee =
+            tf.chainstate.get_chain_config().token_min_supply_change_fee();
+        let token_min_issuance_fee = tf.chainstate.get_chain_config().token_min_issuance_fee();
+
+        let amount_to_mint = Amount::from_atoms(rng.gen_range(1..=i128::MAX as u128));
+        let (token_id, _, utxo_with_change) =
+            issue_token_from_genesis(&mut rng, &mut tf, TokenTotalSupply::Unlimited);
+
+        // Mint more than i128::MAX
+        let mint_tx = TransactionBuilder::new()
+            .add_input(
+                TxInput::from_account(
+                    AccountNonce::new(0),
+                    AccountSpending::TokenTotalSupply(
+                        token_id,
+                        Amount::from_atoms(i128::MAX as u128 + 1),
```

Comment:
It would also be great to have a test where we mint i128::MAX, then mint 1, and see what happens.